src/layers/transformer.rs

```diff
@@ -0,0 +1,177 @@
+///
+/// This is a generic transformer embedding layer. The layer always has piece
+/// embeddings and can optionally have position embeddings, type embeddings,
+/// and a projection of embeddings to the model's hidden size.
+pub struct TransformerEmbeddings {
+    piece_embeddings: Embedding,
+    type_embeddings: Option<Embedding>,
+    position_embeddings: Option<Embedding>,
+    projection: Option<Embedding>,
+    dropouts: EmbeddingLayerDropouts,
+    layer_norms: EmbeddingLayerNorms,
+}
+
+impl TransformerEmbeddings {
+    /// Construct an embeddings layer.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        vb: VarBuilder,
+        dropouts: EmbeddingLayerDropouts,
+        embedding_width: usize,
+        hidden_width: usize,
+        layer_norms: EmbeddingLayerNorms,
+        n_pieces: usize,
+        n_positions: Option<usize>,
+        n_types: Option<usize>,
+    ) -> Result<Self> {
+        let piece_embeddings = embedding(
+            n_pieces,
+            embedding_width,
+            vb.push_prefix("piece_embeddings"),
+        )?;
+
+        let type_embeddings = n_types
+            .map(|n_types| embedding(n_types, embedding_width, vb.push_prefix("type_embeddings")))
+            .transpose()?;
+
+        let position_embeddings = n_positions
+            .map(|n_positions| {
+                embedding(
+                    n_positions,
+                    embedding_width,
+                    vb.push_prefix("position_embeddings"),
+                )
+            })
+            .transpose()?;
+
+        let projection = if embedding_width != hidden_width {
+            Some(embedding(
+                embedding_width,
+                hidden_width,
+                vb.push_prefix("projection"),
+            )?)
+        } else {
+            None
+        };
+
+        Ok(TransformerEmbeddings {
+            dropouts,
+            layer_norms,
+            piece_embeddings,
+            position_embeddings,
+            projection,
+            type_embeddings,
+        })
+    }
+
+    /// Get position identifiers _[0..seq_len)_.
+    fn get_positions(x: &Tensor) -> Result<Tensor> {
+        let (_, seq_len) = x.shape().dims2()?;
+        Ok(Tensor::arange(0, seq_len as i64, x.device())?.reshape((1, seq_len))?)
+    }
+
+    /// Get all-zero type identifiers for the given tensor.
+    fn get_type_ids(x: &Tensor) -> Result<Tensor> {
+        Ok(x.zeros_like()?)
+    }
+
+    /// Calculate the piece embeddings.
+    pub fn forward(
+        &self,
+        piece_ids: &Tensor,
+        positions: Option<&Tensor>,
+        type_ids: Option<&Tensor>,
+    ) -> Result<Tensor> {
+        let mut embeddings = self.piece_embeddings.forward(piece_ids)?;
+
+        if let Some(type_embeddings) = &self.type_embeddings {
+            let type_ids = match type_ids {
+                Some(type_ids) => type_ids.clone(),
+                None => Self::get_type_ids(piece_ids)?,
+            };
```

Comment:
Replaced with `unwrap_or_else` (I think we should avoid bare `unwrap`s if possible). With regards to the `clone`, we have to make the types compatible. The alternative I see is using `Cow`, however that adds much more noise to the code and `clone()` here is just an atomic refcount bump (`Tensor` is `Tensor(Arc<Tensor>)`.

Reply:
Oh no, `or` and `unwrap_or_else` will be more ugly, since `get_type_ids` will return a `Result`, and we cannot transpose the `Option<Result<_>>` unless we also wrap `type_ids` in a `Result`. I think in this case the `match` is just cleaner.